shell/packages/sandstorm-ui-powerbox/powerbox-server.js

```diff
@@ -31,8 +75,240 @@ Meteor.methods({
+  },
+
+  "12214421258504904768"(db, userId, value) {  // IpNetwork
+    if (Meteor.users.findOne(userId).isAdmin) {
+      return [
+        new PowerboxOption({
+          _id: "frontendref-ipnetwork",
+          frontendRef: { ipNetwork: true },
+        }),
+      ];
+    } else {
+      return [];
+    }
   },
+
+  "16369547182874744570"(db, userId, value) {  // IpInterface
+    if (Meteor.users.findOne(userId).isAdmin) {
+      return [
+        new PowerboxOption({
+          _id: "frontendref-ipinterface",
+          frontendRef: { ipInterface: true },
+        }),
+      ];
+    } else {
+      return [];
+    }
+  },
+};
+
+Meteor.publish("powerboxOptions", function (requestId, descriptorList) {
+  // Performs a powerbox query, returning options to present to the user in the powerbox.
+  // `descriptorList` is an array of `PowerboxDescriptor`s, each individually serialized in
+  // packed format and base64-encoded. The publish populates a pseudo-collection called
+  // `powerboxOptions`. Each item has the following fields:
+  //
+  //   _id: Object with two fields: `requestId` (as passed when subscribing) and `optionId` (a
+  //       unique identifier for the option).
+  //   matchQuality: "preferred" or "acceptable" ("unacceptable" options aren't returned).
+  //   frontendRef: If present, selecting this option means creating a simple frontendRef. The
+  //       field value should be passed back to the method `newFrontendRef` verbatim. The format
+  //       of the field is the same as ApiTokens.frontendRef.
+  //   grainId: If present, this option selects a grain to satisfy the request. One or both of
+  //       `uiView` and `hostedObject` will be present. Mutually exclusive with `frontendRef`.
+  //   uiView: If present, this option creates a UiView for a grain. The UI should allow the user
+  //       to choose a role to grant. Only present when `grainId` is also present. `uiView` is an
+  //       object with the fields:
+  //     TODO(someday): Some sort of indication of what options the powerbox should give the user
+  //         as far as which permissions to grant? Unclear if any app would ever want to specify
+  //         permissions when requesting a plain UiView.
+  //   hostedObject: If present, this grain advertises that it publishes capabilities that might
+  //       match the query. If selected, a request session to the grain embedded should be shown
+  //       embedded in the powerbox and the same request should be passed to it. Only present when
+  //       `grainId` is also present. `hostedObject`, when present, is (for now) an empty object.
+
+  const results = {};
+  const db = this.connection.sandstormDb;
+
+  if (descriptorList.length > 0) {
+    const descriptorMatches = descriptorList.map(packedDescriptor => {
+      // Decode the descriptor.
+      // TODO(now): Also single-segment? Canonical?
+
+      // Use URL-safe base64.
+      unUrlsafe = packedDescriptor.replace("-", "+").replace("_", "/");
```

Comment:
Turns out Node documentedly accepts either character set when decoding, so I removed the replacements altogether.